### PR TITLE
[R4R] Improve performance

### PR DESCRIPTION
**Pull request by rickyyangz**
1. Keep nodes in memory after `SaveBranch` and implement in-memory tree pruning
2. disable node cache.
2. preallocate memory for node serialization when SaveNode
3. pre-allocate memory for collecting orphan nodes

### CODE REVIEW

File: mutable_tree.go [hunk 1]
```diff
@@ -8,6 +8,11 @@ import (
 	dbm "github.com/tendermint/tendermint/libs/db"
 )
 
+const (
+	maxVersions = 1000000
+	maxNodes    = 500000
```
Comment (rickyyangz) [inline mutable_tree.go]: will try to make them configurable

File: node_versions.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+func (nv *NodeVersions) getVersion(idx int) int64 {
+	if idx >= nv.nextVersionIdx {
+		return nv.nextVersion - int64(nv.maxVersions) + int64(idx - nv.nextVersionIdx)
+	}
+	return nv.nextVersion - int64(nv.nextVersionIdx - idx)
+}
+
+type Int64Ring struct {
```
Comment (yutianwu) [inline node_versions.go]: delete this

File: node.go [hunk 3]
```diff
@@ -346,32 +365,79 @@ func (node *Node) writeBytes(w io.Writer) cmn.Error {
...
+func (node *Node) getLeftNode(t *ImmutableTree, updateVersion bool) *Node {
+	if node.leftNode == nil {
+		node.mtx.Lock()
+		defer node.mtx.Unlock()
+		if node.leftNode == nil {
+			node.leftNode = t.ndb.GetNode(node.leftHash)
+			if updateVersion {
+				node.leftNode.loadVersion = t.version
+			} else {
+				node.leftNode.loadVersion = node.leftNode.version - 1
```
Comment (yutianwu) [inline node.go]: why minus 1 here

File: node.go [hunk 4]
```diff
@@ -6,11 +6,11 @@ package iavl
 import (
 	"bytes"
 	"fmt"
-	"io"
-
 	"github.com/tendermint/go-amino"
 	"github.com/tendermint/tendermint/crypto/tmhash"
 	cmn "github.com/tendermint/tendermint/libs/common"
+	"io"
+	"sync"
```
Comment (yutianwu) [inline node.go]: fix imports here

File: node.go (hunk 3 above)
Comment (yutianwu) [reply node.go]: as discussed, we can add more comments about loadVersion

File: mutable_tree.go [hunk 5]
```diff
@@ -255,17 +288,23 @@ func (tree *MutableTree) LoadVersion(targetVersion int64) (int64, error) {
...
+		t.root.loadVersion = latestVersion
+		t.nodeVersions.Inc1(t.root.loadVersion)
+		t.isNotEmpty = true
 	}
 
 	tree.orphans = map[string]int64{}
 	tree.ImmutableTree = t
-	tree.lastSaved = t.clone()
+	tree.lastSavedRoot = t.root
+	tree.nodeVersions = nodeVersions
```
Comment (guagualvcha) [inline mutable_tree.go]: this is reduntant.

File: mutable_tree.go [hunk 6]
```diff
@@ -255,17 +288,23 @@ func (tree *MutableTree) LoadVersion(targetVersion int64) (int64, error) {
...
 		t.root = tree.ndb.GetNode(latestRoot)
+		t.root.loadVersion = latestVersion
+		t.nodeVersions.Inc1(t.root.loadVersion)
+		t.isNotEmpty = true
 	}
 
 	tree.orphans = map[string]int64{}
 	tree.ImmutableTree = t
-	tree.lastSaved = t.clone()
+	tree.lastSavedRoot = t.root
```
Comment (guagualvcha) [inline mutable_tree.go]: more clean to assign `lastSavedRoot` to `t` rather than `tree`

File: mutable_tree.go [hunk 7]
```diff
@@ -3,41 +3,50 @@ package iavl
...
 		versions:      map[int64]bool{},
-		ndb:           ndb,
 	}
 }
 
 // IsEmpty returns whether or not the tree has any keys. Only trees that are
 // not empty can be saved.
 func (tree *MutableTree) IsEmpty() bool {
-	return tree.ImmutableTree.Size() == 0
+	return !tree.ImmutableTree.isNotEmpty
```
Comment (guagualvcha) [inline mutable_tree.go]: small tip: `return !tree.isNotEmpty`

File: nodedb.go [hunk 8]
```diff
@@ -107,15 +107,16 @@ func (ndb *nodeDB) SaveNode(node *Node) {
...
+	buf.Grow(node.aminoSize())
+	if err := node.writeBytes(&buf); err != nil {
 		panic(err)
 	}
 	ndb.batch.Set(ndb.nodeKey(node.hash), buf.Bytes())
 	debug("BATCH SAVE %X %p\n", node.hash, node)
 
 	node.persisted = true
-	ndb.cacheNode(node)
+	// ndb.cacheNode(node)
```
Comment (yutianwu) [inline nodedb.go]: remember to delete these comments too

File: mutable_tree.go [hunk 9]
```diff
@@ -448,51 +548,49 @@ func (tree *MutableTree) rotateLeft(node *Node) (*Node, *Node) {
...
-			return newNode, []*Node{orphaned}
+			newNode := tree.rotateRight(node, orphans)
+			return newNode
 		}
 		// Left Right Case
-		var leftOrphaned *Node
-
-		left := node.getLeftNode(tree.ImmutableTree)
+		*orphans = append(*orphans, left)
+		// tree.nodeVersions.Dec1(left.version)
```
Comment (yutianwu) [inline mutable_tree.go]: as discussed, we can move orphan collecting into rotate function

File: node.go [hunk 10]
```diff
@@ -26,16 +26,20 @@ type Node struct {
...
-		value:   value,
-		height:  0,
-		size:    1,
-		version: version,
+		key:         key,
+		value:       value,
+		height:      0,
+		size:        1,
+		version:     version,
+		loadVersion: version - 1,
```
Comment (yutianwu) [inline node.go]: better add comments here to explain why we minus 1 here

File: node_versions.go [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+func (nv *NodeVersions) Inc1(version int64) {
+	nv.changes[version]++
+}
+
+func (nv *NodeVersions) Inc(version int64, times int) {
+	nv.changes[version] += times
+}
+
+func (nv *NodeVersions) Dec1(version int64) {
+	nv.changes[version]--
```
Comment (yutianwu) [inline node_versions.go]: should we check nv.changes[version] > 0 before we minus one

File: mutable_tree.go [hunk 12]
```diff
@@ -64,76 +73,87 @@ func (tree *MutableTree) String() string {
...
-	newSelf *Node, updated bool, orphaned []*Node,
+func (tree *MutableTree) recursiveSet(node *Node, key []byte, value []byte, orphans *[]*Node) (
+	newSelf *Node, updated bool,
 ) {
 	version := tree.version + 1
 
 	if node.isLeaf() {
 		switch bytes.Compare(key, node.key) {
 		case -1:
+			tree.nodeVersions.Inc(tree.version, 2)
```
Comment (yutianwu) [inline mutable_tree.go]: should we minus 1 old version here

File: node_versions.go [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+		nextVersionIdx: 0,
+		totalNodes:     0,
+	}
+}
+
+func (nv *NodeVersions) Inc1(version int64) {
+	nv.changes[version]++
+}
+
+func (nv *NodeVersions) Inc(version int64, times int) {
```
[the remark this answers is not in the document]
Comment (rickyyangz) [reply node_versions.go]: we need a common method `Inc` and `Dec`. Inc1 and Dec1 are just convenient methods

File: mutable_tree.go [hunk 14]
```diff
@@ -365,10 +393,42 @@ func (tree *MutableTree) SaveVersion() ([]byte, int64, error) {
...
+		return nil, version, err
+	}
+	if maxPruneVersion > 0 {
+		tree.PruneInMemory(maxPruneVersion)
+	}
 	return tree.Hash(), version, nil
 }
 
+func (tree *MutableTree) PruneInMemory(maxPruneVersion int64) {
+	var iter func(root *Node)
```
[the remark this answers is not in the document]
Comment (rickyyangz) [reply mutable_tree.go]: it's an inner function, if you want to invoke it recursively, you need to define it first

File: node.go [hunk 15]
```diff
@@ -299,6 +299,20 @@ func (node *Node) writeHashBytesRecursively(w io.Writer) (hashCount int64, err c
 	return
 }
 
+func (node *Node) aminoSize() int {
+	n := 1 +
```
[the remark this answers is not in the document]
Comment (rickyyangz) [reply node.go]: added comments

File: node_versions.go (hunk 11 above)
Comment (rickyyangz) [reply node_versions.go]: no, the changes can be negative

File: node.go (hunk 3 above)
Comment (rickyyangz) [reply node.go]: added

**Pull request merged**